shared/fs/footer/upload-countdown-hoc.js

```diff
@@ -0,0 +1,160 @@
+// @flow
+import * as React from 'react'
+import {compose} from '../../util/container'
+import {formatDuration} from '../../util/timestamp'
+import {HOCTimers, type PropsWithTimer} from '../../common-adapters'
+import {type UploadProps} from './upload'
+
+export type UploadCountdownHOCProps = {
+  endEstimate?: number,
+  files: number,
+  debugToggleShow?: () => void,
+}
+
+type Props = PropsWithTimer<UploadCountdownHOCProps>
+
+// Cosider this component as a state machine with following four states. 1Hz
+// Ticks (from _tick() calls by setInterval) and props changes (through
+// componentDidUpdate() calls) are two possible inputs.
+type Mode =
+  // The upload banner isn't shown.
+  | 'hidden'
+  // Normal count-down. If upload is finished during this state while glueTTL
+  // is smaller than or equal to 0, transition to hidden. If upload is finished
+  // during this state while glueTTL is greater than 0, transition to sticky.
+  | 'count-down'
+  // The upload banner should have been hidden but we are still showing it
+  // because it hasn't been showed for long enough. When glueTTL hits 0,
+  // transition to 0.
+  | 'sticky'
+
+type State = {
+  displayDuration: number,
+  mode: Mode,
+  glueTTL: number,
+}
+
+const tickInterval = 1000
+const initialGlueTTL = 2
+
+const UploadCountdownHOC = (Upload: React.ComponentType<UploadProps>) =>
+  class extends React.PureComponent<Props, State> {
+    state = {
+      displayDuration: 0,
+      mode: 'hidden',
+      glueTTL: 0,
+    }
+
+    _tickerID: ?IntervalID = null
+
+    _tick = () => {
+      const {mode, glueTTL, displayDuration} = this.state
+      const newDisplayDuration = displayDuration > 1000 ? displayDuration - 1000 : 0
+      const newGlueTTL = glueTTL > 1 ? glueTTL - 1 : 0
+      switch (mode) {
+        case 'hidden':
+          this._stopTicker()
+          break
+        case 'count-down':
+          this.setState({
+            displayDuration: newDisplayDuration,
+            glueTTL: newGlueTTL,
+          })
+          break
+        case 'sticky':
+          this.setState({
+            mode: newGlueTTL > 0 ? 'sticky' : 'hidden',
+            glueTTL: newGlueTTL,
+            displayDuration: newDisplayDuration,
+          })
+          break
+        default:
+          /*::
+      declare var ifFlowErrorsHereItsCauseYouDidntHandleAllActionTypesAbove: (mode: empty) => any
+      ifFlowErrorsHereItsCauseYouDidntHandleAllActionTypesAbove(mode);
+      */
+          break
+      }
+    }
+
+    _startTicker = () => {
+      if (this._tickerID) {
+        return
+      }
+      this._tickerID = this.props.setInterval(this._tick, tickInterval)
```

Comment:
likely want to call this._stopTicker here

Reply:
I meant this call to be idempotent, i.e. start the ticker if it hasn't been started. The `if` condition above already covers the case where we already have a ticker ID. I'll add a comment to clarify.